homeassistant/components/modbus/__init__.py

```diff
@@ -88,163 +97,219 @@
 )
 
 
-def setup_client(client_config):
-    """Set up pymodbus client."""
-    client_type = client_config[CONF_TYPE]
-
-    if client_type == "serial":
-        return ModbusSerialClient(
-            method=client_config[CONF_METHOD],
-            port=client_config[CONF_PORT],
-            baudrate=client_config[CONF_BAUDRATE],
-            stopbits=client_config[CONF_STOPBITS],
-            bytesize=client_config[CONF_BYTESIZE],
-            parity=client_config[CONF_PARITY],
-            timeout=client_config[CONF_TIMEOUT],
-        )
-    if client_type == "rtuovertcp":
-        return ModbusTcpClient(
-            host=client_config[CONF_HOST],
-            port=client_config[CONF_PORT],
-            framer=ModbusRtuFramer,
-            timeout=client_config[CONF_TIMEOUT],
-        )
-    if client_type == "tcp":
-        return ModbusTcpClient(
-            host=client_config[CONF_HOST],
-            port=client_config[CONF_PORT],
-            timeout=client_config[CONF_TIMEOUT],
-        )
-    if client_type == "udp":
-        return ModbusUdpClient(
-            host=client_config[CONF_HOST],
-            port=client_config[CONF_PORT],
-            timeout=client_config[CONF_TIMEOUT],
-        )
-    assert False
-
-
-def setup(hass, config):
+async def async_setup(hass, config):
     """Set up Modbus component."""
-    hass.data[DOMAIN] = hub_collect = {}
+    hass.data[MODBUS_DOMAIN] = hub_collect = {}
 
-    for client_config in config[DOMAIN]:
-        client = setup_client(client_config)
-        name = client_config[CONF_NAME]
-        hub_collect[name] = ModbusHub(client, name)
-        _LOGGER.debug("Setting up hub: %s", client_config)
+    _LOGGER.debug("registering hubs")
+    for client_config in config[MODBUS_DOMAIN]:
+        hub_collect[client_config[CONF_NAME]] = ModbusHub(client_config, hass.loop)
 
     def stop_modbus(event):
         """Stop Modbus service."""
         for client in hub_collect.values():
-            client.close()
+            del client
```

Comment:
Is this going to shutdown cleanly? 

Does the underlying protocol need to be stopped here?

Reply:
yes, there are no longer a .close() method. Pymodbus catches the del and does whatever needs to be done.

Need to, is a matter of taste, since hass is terminating (at least that is the only way I could provoke the stop event) everything will be shutdown automatically. However maybe we would want to be able to shutdown a single integration and then it is needed.